<a href="https://colab.research.google.com/github/pdesire-20/Lab6/blob/main/lab_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Welcome to lab_🐼s!

# Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In every lab, you will need to import the library just like we did in lecture!

In the following cell, replace `...` with the code to import pandas as `pd`:

In [3]:
# Write the code to import pandas as pd:
import pandas as pd


In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```

Remember, URLs passed to `read_csv` are **string** variables, so remember to use **quotation marks**.

In [7]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog.csv")


In [8]:
# Look at `df` to confirm that this worked:
df

Year    Term YearTerm Subject  Number                          Name  \
0      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
1      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
2      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
3      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
4      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
...     ...     ...      ...     ...     ...                           ...   
11898  2024  Spring  2024-sp    WRIT     303          Writing Across Media   
11899  2024  Spring  2024-sp    WRIT     303          Writing Across Media   
11900  2024  Spring  2024-sp    YDSH     102          Beginning Yiddish II   
11901  2024  Spring  2024-sp    YDSH     102          Beginning Yiddish II   
11902  2024  Spring  2024-sp    ZULU     202            Elementary Zulu II   

                                             Description Credit Hours  \
0      Interdisciplinary introduction to the basic co...     3 hours.   
1      Interdisciplinary introduction to the basic co...     3 hours.   
2      Interdisciplinary introduction to the basic co...     3 hours.   
3      Interdisciplinary introduction to the basic co...     3 hours.   
4      Interdisciplinary introduction to the basic co...     3 hours.   
...                                                  ...          ...   
11898                    Same as INFO 303. See INFO 303.     3 hours.   
11899                    Same as INFO 303. See INFO 303.     3 hours.   
11900  Continuation of YDSH 101 focusing on comprehen...     4 hours.   
11901  Continuation of YDSH 101 focusing on comprehen...     4 hours.   
11902  Continuation of ZULU 201 with introduction of ...     5 hours.   

                                            Section Info  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
11898                    Same as INFO 303. See INFO 303.   
11899                    Same as INFO 303. See INFO 303.   
11900                                                NaN   
11901                                                NaN   
11902  Same as AFST 252. Participation in the languag...   

                                       Degree Attributes  ... Section Status  \
0      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
...                                                  ...  ...            ...   
11898                       Advanced Composition course.  ...              A   
11899                       Advanced Composition course.  ...              A   
11900                                                NaN  ...              A   
11901                                                NaN  ...              A   
11902                                                NaN  ...              A   

                Enrollment Status                   Type Type Code Start Time  \
0                            Open  Discussion/Recitation       DIS   09:00 AM   
1                            Open  Discussion/Recitation       DIS   10:00 AM   
2                            Open  Discussion/Recitation       DIS   11:00 AM   
3                            Open  Discussion/Recitation       DIS   12:00 PM   
4                            Open  Discussion/Recitation       DIS   01:00 PM   
...                      

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [10]:
## == CHECKPOINT TEST CASES for Part 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 11903 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [14]:
df_STAT = df[df["Subject"] == "STAT"]
df_STAT

Year    Term YearTerm Subject  Number                           Name  \
11168  2024  Spring  2024-sp    STAT     100                     Statistics   
11169  2024  Spring  2024-sp    STAT     100                     Statistics   
11170  2024  Spring  2024-sp    STAT     100                     Statistics   
11171  2024  Spring  2024-sp    STAT     107         Data Science Discovery   
11172  2024  Spring  2024-sp    STAT     107         Data Science Discovery   
...     ...     ...      ...     ...     ...                            ...   
11312  2024  Spring  2024-sp    STAT     587     Hierarchical Linear Models   
11313  2024  Spring  2024-sp    STAT     590  Individual Study and Research   
11314  2024  Spring  2024-sp    STAT     593                STAT Internship   
11315  2024  Spring  2024-sp    STAT     593                STAT Internship   
11316  2024  Spring  2024-sp    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11168  First course in probability and statistics at ...        3 hours.   
11169  First course in probability and statistics at ...        3 hours.   
11170  First course in probability and statistics at ...        3 hours.   
11171  Data Science Discovery is the intersection of ...        4 hours.   
11172  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
11312       Same as PSYC 587 and EPSY 587. See EPSY 587.        4 hours.   
11313  Directed reading and research. Approved for le...   0 TO 8 hours.   
11314  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11315  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11316  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11168  Credit is not given for both STAT 100 and any ...   
11169  Credit is not given for both STAT 100 and any ...   
11170  Credit is not given for both STAT 100 and any ...   
11171                         Same as CS 107 and IS 107.   
11172                         Same as CS 107 and IS 107.   
...                                                  ...   
11312       Same as PSYC 587 and EPSY 587. See EPSY 587.   
11313  Approved for letter and S/U grading. May be re...   
11314  Approved for letter and S/U grading. Prerequis...   
11315  Approved for letter and S/U grading. Prerequis...   
11316  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11168  Quantitative Reasoning I course.  ...              A   
11169  Quantitative Reasoning I course.  ...              A   
11170  Quantitative Reasoning I course.  ...              A   
11171  Quantitative Reasoning I course.  ...              A   
11172  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
11312                               NaN  ...              A   
11313                               NaN  ...              A   
11314                               NaN  ...              A   
11315                               NaN  ...              A   
11316                               NaN  ...              A   

                Enrollment Status                   Type Type Code Start Time  \
11168                        Open                Lecture       LEC   11:00 AM   
11169                      Closed                Lecture       LEC   12:00 PM   
11170           Open (Restricted)                Lecture       LEC   02:00 PM   
11171                      Closed                Lecture       LEC   10:00 AM   
11172                      Closed  Laboratory-Discussion       LBD   01:00 PM   
...                           ...                    ...       ...        ...   
11312  CrossListOpen (Restricted)                Lecture       LEC   10:00 AM   
11313              

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [15]:
df_CS = df[df["Subject"] == "CS"]
df_CS

Year    Term YearTerm Subject  Number                              Name  \
3026  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3027  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3028  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3029  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3030  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...     ...      ...     ...     ...                               ...   
3402  2024  Spring  2024-sp      CS     598                    Special Topics   
3403  2024  Spring  2024-sp      CS     598                    Special Topics   
3404  2024  Spring  2024-sp      CS     598                    Special Topics   
3405  2024  Spring  2024-sp      CS     598                    Special Topics   
3406  2024  Spring  2024-sp      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3026  Fundamental principles, concepts, and methods ...        3 hours.   
3027  Fundamental principles, concepts, and methods ...        3 hours.   
3028  Fundamental principles, concepts, and methods ...        3 hours.   
3029  Fundamental principles, concepts, and methods ...        3 hours.   
3030  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3402  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3403  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3404  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3405  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3406    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3026  Prerequisite: One of MATH 220 or MATH 221 or M...   
3027  Prerequisite: One of MATH 220 or MATH 221 or M...   
3028  Prerequisite: One of MATH 220 or MATH 221 or M...   
3029  Prerequisite: One of MATH 220 or MATH 221 or M...   
3030  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3402  May be repeated in the same or separate terms ...   
3403  May be repeated in the same or separate terms ...   
3404  May be repeated in the same or separate terms ...   
3405  May be repeated in the same or separate terms ...   
3406    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3026  Quantitative Reasoning II course.  ...              A   
3027  Quantitative Reasoning II course.  ...              A   
3028  Quantitative Reasoning II course.  ...              A   
3029  Quantitative Reasoning II course.  ...              A   
3030  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3402                                NaN  ...              A   
3403                                NaN  ...              A   
3404                                NaN  ...              A   
3405                                NaN  ...              A   
3406                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3026  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3027             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3028             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3029             Closed  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
3030               Open  Laboratory-Discussion       LBD   11:00 AM  12:50 PM   
...                 ...                    ...       ...        ...       ...   
3402  Open (Restricted)     Lecture-Discussion        S4   02:00 PM  03:15 PM   
3403             C

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [16]:
df_IS = df[df["Subject"] == "IS"]
df_IS

Year    Term YearTerm Subject  Number  \
6090  2024  Spring  2024-sp      IS     100   
6091  2024  Spring  2024-sp      IS     101   
6092  2024  Spring  2024-sp      IS     101   
6093  2024  Spring  2024-sp      IS     101   
6094  2024  Spring  2024-sp      IS     101   
...    ...     ...      ...     ...     ...   
6346  2024  Spring  2024-sp      IS     597   
6347  2024  Spring  2024-sp      IS     597   
6348  2024  Spring  2024-sp      IS     597   
6349  2024  Spring  2024-sp      IS     597   
6350  2024  Spring  2024-sp      IS     599   

                                                   Name  \
6090   Exploring the iSchool with a Human-Centered Lens   
6091               Introduction to Information Sciences   
6092               Introduction to Information Sciences   
6093               Introduction to Information Sciences   
6094               Introduction to Information Sciences   
...                                                 ...   
6346  Advanced Topics in Data Analytics &amp; Data S...   
6347  Advanced Topics in Data Analytics &amp; Data S...   
6348  Advanced Topics in Data Analytics &amp; Data S...   
6349  Advanced Topics in Data Analytics &amp; Data S...   
6350                                    Thesis Research   

                                            Description    Credit Hours  \
6090  This course introduces students to the School ...        1 hours.   
6091  This course provides an introduction to the fi...        3 hours.   
6092  This course provides an introduction to the fi...        3 hours.   
6093  This course provides an introduction to the fi...        3 hours.   
6094  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6346  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6347  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6348  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6349  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6350  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6090  Prerequisite: Restricted to Majors Only; First...   
6091                                                NaN   
6092                                                NaN   
6093                                                NaN   
6094                                                NaN   
...                                                 ...   
6346  2 or 4 graduate hours. No professional credit....   
6347  2 or 4 graduate hours. No professional credit....   
6348  2 or 4 graduate hours. No professional credit....   
6349  2 or 4 graduate hours. No professional credit....   
6350  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6090                                 NaN  ...              A   
6091  Social & Beh Sci - Soc Sci course.  ...              A   
6092  Social & Beh Sci - Soc Sci course.  ...              A   
6093  Social & Beh Sci - Soc Sci course.  ...              A   
6094  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6346                                 NaN  ...              A   
6347                                 NaN  ...              A   
6348                                 NaN  ...              A   
6349                                 NaN  ...              A   
6350                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6090  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6091               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6092               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6093               Open  Disc

### 🔬 Test Case Checkpoint 🔬


In [17]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_CS to only CS courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_IS to only IS courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?


:*(✏️ STAT has 141 CS 381. ✏️)*

In [26]:
len(df_STAT), len(df_CS), len(df_IS)


(149, 381, 261)

### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest numbered course?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


*(✏️ There are only 4 introductory classes for all of them. ✏️)*

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- DPI, feel free to pick a major you're interested in exploring.
- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!

### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [28]:
mydf = df[df["Subject"]=="CS"]
mydf

Year    Term YearTerm Subject  Number                              Name  \
3026  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3027  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3028  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3029  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3030  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...     ...      ...     ...     ...                               ...   
3402  2024  Spring  2024-sp      CS     598                    Special Topics   
3403  2024  Spring  2024-sp      CS     598                    Special Topics   
3404  2024  Spring  2024-sp      CS     598                    Special Topics   
3405  2024  Spring  2024-sp      CS     598                    Special Topics   
3406  2024  Spring  2024-sp      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3026  Fundamental principles, concepts, and methods ...        3 hours.   
3027  Fundamental principles, concepts, and methods ...        3 hours.   
3028  Fundamental principles, concepts, and methods ...        3 hours.   
3029  Fundamental principles, concepts, and methods ...        3 hours.   
3030  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3402  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3403  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3404  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3405  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3406    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3026  Prerequisite: One of MATH 220 or MATH 221 or M...   
3027  Prerequisite: One of MATH 220 or MATH 221 or M...   
3028  Prerequisite: One of MATH 220 or MATH 221 or M...   
3029  Prerequisite: One of MATH 220 or MATH 221 or M...   
3030  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3402  May be repeated in the same or separate terms ...   
3403  May be repeated in the same or separate terms ...   
3404  May be repeated in the same or separate terms ...   
3405  May be repeated in the same or separate terms ...   
3406    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3026  Quantitative Reasoning II course.  ...              A   
3027  Quantitative Reasoning II course.  ...              A   
3028  Quantitative Reasoning II course.  ...              A   
3029  Quantitative Reasoning II course.  ...              A   
3030  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3402                                NaN  ...              A   
3403                                NaN  ...              A   
3404                                NaN  ...              A   
3405                                NaN  ...              A   
3406                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3026  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3027             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3028             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3029             Closed  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
3030               Open  Laboratory-Discussion       LBD   11:00 AM  12:50 PM   
...                 ...                    ...       ...        ...       ...   
3402  Open (Restricted)     Lecture-Discussion        S4   02:00 PM  03:15 PM   
3403             C

In [27]:
df.Subject.unique()

array(['AAS', 'ABE', 'ACCY', 'ACE', 'ACES', 'ADV', 'AE', 'AFAS', 'AFRO',
       'AFST', 'AGCM', 'AGED', 'AHS', 'AIS', 'ALEC', 'ANSC', 'ANTH',
       'ARAB', 'ARCH', 'ART', 'ARTD', 'ARTE', 'ARTF', 'ARTH', 'ARTJ',
       'ARTS', 'ASRM', 'ASST', 'ASTR', 'ATMS', 'BADM', 'BASQ', 'BCOG',
       'BCS', 'BDI', 'BIOC', 'BIOE', 'BIOP', 'BSE', 'BTW', 'BUS', 'CAS',
       'CB', 'CDB', 'CEE', 'CHBE', 'CHEM', 'CHIN', 'CHLH', 'CHP', 'CI',
       'CIC', 'CLCV', 'CLE', 'CMN', 'CPSC', 'CS', 'CSE', 'CW', 'CWL',
       'CZCH', 'DANC', 'DTX', 'EALC', 'ECE', 'ECON', 'EDPR', 'EDUC',
       'EEB', 'EIL', 'ENG', 'ENGL', 'ENSU', 'ENT', 'ENVS', 'EPOL', 'EPSY',
       'ERAM', 'ESE', 'ESL', 'ETMA', 'EURO', 'FAA', 'FIN', 'FLTE', 'FR',
       'FSHN', 'GC', 'GEOL', 'GER', 'GGIS', 'GLBL', 'GMC', 'GRK', 'GRKM',
       'GS', 'GSD', 'GWS', 'HDFS', 'HEBR', 'HIST', 'HNDI', 'HORT', 'HT',
       'HUM', 'IB', 'IE', 'IHLT', 'INFO', 'IS', 'ITAL', 'JAPN', 'JOUR',
       'JS', 'KIN', 'KOR', 'LA', 'LAS', 'LAST', 'LAT', 'LAW', 'LCT

### Puzzle 4.2:

Select **five random course sections** from your major, using `mydf` and pandas functions:

In [29]:
mydf.sample(n=5)

Year    Term YearTerm Subject  Number                              Name  \
3317  2024  Spring  2024-sp      CS     483      Applied Parallel Programming   
3139  2024  Spring  2024-sp      CS     225                   Data Structures   
3136  2024  Spring  2024-sp      CS     211  Ethical and Professional Conduct   
3075  2024  Spring  2024-sp      CS     107            Data Science Discovery   
3030  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   

                                            Description Credit Hours  \
3317          Same as CSE 408 and ECE 408. See ECE 408.     4 hours.   
3139  Data abstractions: elementary data structures ...     4 hours.   
3136  Navigating the complex ethical and professiona...     3 hours.   
3075         Same as IS 107 and STAT 107. See STAT 107.     4 hours.   
3030  Fundamental principles, concepts, and methods ...     3 hours.   

                                           Section Info  \
3317          Same as CSE 408 and ECE 408. See ECE 408.   
3139  Credit is not given for CS 277 if credit for C...   
3136  Credit is not given for both CS 211 and CS 210...   
3075         Same as IS 107 and STAT 107. See STAT 107.   
3030  Prerequisite: One of MATH 220 or MATH 221 or M...   

                      Degree Attributes  ... Section Status  \
3317                                NaN  ...              A   
3139  Quantitative Reasoning II course.  ...              A   
3136       Advanced Composition course.  ...              A   
3075   Quantitative Reasoning I course.  ...              A   
3030  Quantitative Reasoning II course.  ...              A   

               Enrollment Status                   Type Type Code Start Time  \
3317                      Closed         Online Lecture       OLC   09:30 AM   
3139  CrossListOpen (Restricted)             Laboratory       LAB   05:00 PM   
3136           Open (Restricted)                 Online       ONL   10:00 AM   
3075                      Closed  Laboratory-Discussion       LBD   02:00 PM   
3030                        Open  Laboratory-Discussion       LBD   11:00 AM   

      End Time Days of Week  Room                       Building  \
3317  10:50 AM           TR   NaN                            NaN   
3139  05:50 PM            M   216     Siebel Center for Comp Sci   
3136  12:50 PM            F   NaN                            NaN   
3075  03:20 PM            R   G46  Literatures, Cultures, & Ling   
3030  12:50 PM            T  2036  Campus Instructional Facility   

                 Instructors  
3317          Kindratenko, V  
3139  Erickson, J;Solomon, B  
3136               Winter, G  
3075                     NaN  
3030              Beckman, M  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven lowest numbered course sections** from your major:

In [32]:
mydf.nsmallest(7,'Number' )

Year    Term YearTerm Subject  Number                              Name  \
3026  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3027  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3028  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3029  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3030  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3031  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3032  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   

                                            Description Credit Hours  \
3026  Fundamental principles, concepts, and methods ...     3 hours.   
3027  Fundamental principles, concepts, and methods ...     3 hours.   
3028  Fundamental principles, concepts, and methods ...     3 hours.   
3029  Fundamental principles, concepts, and methods ...     3 hours.   
3030  Fundamental principles, concepts, and methods ...     3 hours.   
3031  Fundamental principles, concepts, and methods ...     3 hours.   
3032  Fundamental principles, concepts, and methods ...     3 hours.   

                                           Section Info  \
3026  Prerequisite: One of MATH 220 or MATH 221 or M...   
3027  Prerequisite: One of MATH 220 or MATH 221 or M...   
3028  Prerequisite: One of MATH 220 or MATH 221 or M...   
3029  Prerequisite: One of MATH 220 or MATH 221 or M...   
3030  Prerequisite: One of MATH 220 or MATH 221 or M...   
3031  Prerequisite: One of MATH 220 or MATH 221 or M...   
3032  Prerequisite: One of MATH 220 or MATH 221 or M...   

                      Degree Attributes  ... Section Status  \
3026  Quantitative Reasoning II course.  ...              A   
3027  Quantitative Reasoning II course.  ...              A   
3028  Quantitative Reasoning II course.  ...              A   
3029  Quantitative Reasoning II course.  ...              A   
3030  Quantitative Reasoning II course.  ...              A   
3031  Quantitative Reasoning II course.  ...              A   
3032  Quantitative Reasoning II course.  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3026  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3027             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3028             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3029             Closed  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
3030               Open  Laboratory-Discussion       LBD   11:00 AM  12:50 PM   
3031             Closed  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3032             Closed  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   

     Days of Week   Room                       Building Instructors  
3026           MW  THEAT                   Lincoln Hall  Beckman, M  
3027            M     18  Campus Instructional Facility  Beckman, M  
3028            M     18  Campus Instructional Facility  Beckman, M  
3029            T   1038  Campus Instructional Facility  Beckman, M  
3030            T   2036  Campus Instructional Facility  Beckman, M  
3031            T    134                         Armory  Beckman, M  
3032            T     18  Campus Instructional Facility  Beckman, M  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine highest numbered course sections** from your major:

In [33]:
mydf.nlargest(9,'Number' )

Year    Term YearTerm Subject  Number             Name  \
3406  2024  Spring  2024-sp      CS     599  Thesis Research   
3389  2024  Spring  2024-sp      CS     598   Special Topics   
3390  2024  Spring  2024-sp      CS     598   Special Topics   
3391  2024  Spring  2024-sp      CS     598   Special Topics   
3392  2024  Spring  2024-sp      CS     598   Special Topics   
3393  2024  Spring  2024-sp      CS     598   Special Topics   
3394  2024  Spring  2024-sp      CS     598   Special Topics   
3395  2024  Spring  2024-sp      CS     598   Special Topics   
3396  2024  Spring  2024-sp      CS     598   Special Topics   

                                            Description    Credit Hours  \
3406    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   
3389  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3390  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3391  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3392  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3393  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3394  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3395  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3396  Subject offerings of new and developing areas ...   2 TO 4 hours.   

                                           Section Info Degree Attributes  \
3406    Approved for S/U grading only. May be repeated.               NaN   
3389  May be repeated in the same or separate terms ...               NaN   
3390  May be repeated in the same or separate terms ...               NaN   
3391  May be repeated in the same or separate terms ...               NaN   
3392  May be repeated in the same or separate terms ...               NaN   
3393  May be repeated in the same or separate terms ...               NaN   
3394  May be repeated in the same or separate terms ...               NaN   
3395  May be repeated in the same or separate terms ...               NaN   
3396  May be repeated in the same or separate terms ...               NaN   

      ... Section Status           Enrollment Status                Type  \
3406  ...              A                      Closed   Independent Study   
3389  ...              A                      Closed  Lecture-Discussion   
3390  ...              A           Open (Restricted)  Lecture-Discussion   
3391  ...              A           Open (Restricted)  Lecture-Discussion   
3392  ...              A                      Closed  Lecture-Discussion   
3393  ...              A                      Closed              Online   
3394  ...              A  CrossListOpen (Restricted)              Online   
3395  ...              A                      Closed  Lecture-Discussion   
3396  ...              A           Open (Restricted)  Lecture-Discussion   

     Type Code Start Time  End Time Days of Week  Room  \
3406       IND   ARRANGED       NaN          NaN   NaN   
3389        S9   02:00 PM  03:15 PM           TR  1214   
3390        S7   11:00 AM  12:15 PM           TR  1302   
3391        S3   11:00 AM  12:15 PM           TR  1214   
3392       S10   02:00 PM  03:15 PM           WF  3217   
3393       E16   ARRANGED       NaN          NaN   NaN   
3394       ONL   ARRANGED       NaN          NaN   NaN   
3395       S13   02:00 PM  03:15 PM           TR  1304   
3396       S12   11:00 AM  12:15 PM           WF  3217   

                        Building     Instructors  
3406                         NaN             NaN  
3389  Siebel Center for Comp Sci        Zhang, M  
3390  Siebel Center for Comp Sci    Kloeckner, A  
3391  Siebel Center for Comp Sci        Heath, D  
3392          Everitt Laboratory  Hakkani Tur, D  
3393                         NaN          Sun, J  
3394                         NaN         Park, T  
3395  Siebel Center for Comp Sci  Soltanaghai, E  
3396          Everitt Laboratory    Chowdh

### Puzzle 4.5:

Select **a random sample of 10% of the course sections** from your major:

In [34]:
mydf.sample(frac=0.1 )

Year    Term YearTerm Subject  Number  \
3304  2024  Spring  2024-sp      CS     473   
3396  2024  Spring  2024-sp      CS     598   
3228  2024  Spring  2024-sp      CS     415   
3155  2024  Spring  2024-sp      CS     307   
3193  2024  Spring  2024-sp      CS     374   
3383  2024  Spring  2024-sp      CS     591   
3244  2024  Spring  2024-sp      CS     433   
3358  2024  Spring  2024-sp      CS     536   
3376  2024  Spring  2024-sp      CS     591   
3393  2024  Spring  2024-sp      CS     598   
3399  2024  Spring  2024-sp      CS     598   
3052  2024  Spring  2024-sp      CS     105   
3091  2024  Spring  2024-sp      CS     124   
3363  2024  Spring  2024-sp      CS     562   
3247  2024  Spring  2024-sp      CS     433   
3090  2024  Spring  2024-sp      CS     124   
3196  2024  Spring  2024-sp      CS     374   
3138  2024  Spring  2024-sp      CS     222   
3375  2024  Spring  2024-sp      CS     591   
3279  2024  Spring  2024-sp      CS     445   
3048  2024  Spring  2024-sp      CS     102   
3027  2024  Spring  2024-sp      CS     101   
3168  2024  Spring  2024-sp      CS     341   
3312  2024  Spring  2024-sp      CS     482   
3284  2024  Spring  2024-sp      CS     448   
3167  2024  Spring  2024-sp      CS     341   
3272  2024  Spring  2024-sp      CS     443   
3226  2024  Spring  2024-sp      CS     415   
3248  2024  Spring  2024-sp      CS     434   
3227  2024  Spring  2024-sp      CS     415   
3125  2024  Spring  2024-sp      CS     199   
3051  2024  Spring  2024-sp      CS     105   
3209  2024  Spring  2024-sp      CS     410   
3136  2024  Spring  2024-sp      CS     211   
3084  2024  Spring  2024-sp      CS     124   
3282  2024  Spring  2024-sp      CS     446   
3183  2024  Spring  2024-sp      CS     374   
3088  2024  Spring  2024-sp      CS     124   

                                                   Name  \
3304                                         Algorithms   
3396                                     Special Topics   
3228                                   Game Development   
3155              Modeling and Learning in Data Science   
3193  Introduction to Algorithms &amp; Models of Com...   
3383                                   Advanced Seminar   
3244                       Computer System Organization   
3358                     Fault-Tolerant Dig Syst Design   
3376                                   Advanced Seminar   
3393                                     Special Topics   
3399                                     Special Topics   
3052                          Intro Computing: Non-Tech   
3091                 Introduction to Computer Science I   
3363  Advanced Topics in Security, Privacy, and Mach...   
3247                       Computer System Organization   
3090                 Introduction to Computer Science I   
3196  Introduction to Algorithms &amp; Models of Com...   
3138                                Software Design Lab   
3375                                   Advanced Seminar   
3279                          Computational Photography   
3048                           Little Bits to Big Ideas   
3027                   Intro Computing: Engrg &amp; Sci   
3168                                 System Programming   
3312                                         Simulation   
3284                         Audio Computing Laboratory   
3167                                 System Programming   
3272                             Reinforcement Learning   
3226                                   Game Development   
3248     Real World Algorithms for IoT and Data Science   
3227                                   Game Development   
3125     Undergraduate Open Seminar in Computer Science   
3051                          Intro Computing: Non-Tech   
3209                           Text Information Systems   
3136                   Ethical and Professional Conduct   
3084                 Introduction to Computer Science I   
3282                                   Machine L

### ✔️ Checkpoint ✔️

There are no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is still a new relatively new major here at UIUC. The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [37]:
df2019 = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the Spring 2024 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [38]:
df2019[df2019["Subject"] == "IS"]

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [ ]:
...

In [ ]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

There is a major incerase from 2019 to 2024. from 92 to 261

<hr style="color: #DD3403;">